fastestimator/pipeline.py

```diff
@@ -13,442 +13,132 @@
 
-T = TypeVar('T')
-
-
-class OpDataset(Dataset):
-    def __init__(self, dataset: Dataset, ops: List[Union[NumpyOp, Scheduler[NumpyOp]]]):
-        self.dataset = dataset
-        self.signature_epochs = get_signature_epochs(ops)
-        self.op_schedule = None if len(ops) == 0 else Scheduler(
-            {epoch: get_per_epoch(ops, epoch)
-             for epoch in self.signature_epochs})
-        # The following variables could technically by computed on the fly, but don't want to waste clock cycles
-        self.epoch, self.epoch_ops, self.state_dict = 0, [], []
-        self.set_epoch_and_mode(0, "train")
-
-    def __getitem__(self, index):
-        item = self.dataset.__getitem__(index)
-        op_data = None
-        for op in self.epoch_ops:
-            op_data = get_inputs_by_op(op, item, op_data)
-            op_data = op.forward(op_data, self.state_dict)
-            if op.outputs:
-                write_outputs_by_key(item, op_data, op.outputs)
-        return item
-
-    def __len__(self):
-        return len(self.dataset)
-
-    def set_epoch_and_mode(self, epoch: int, mode: str):
-        self.epoch = epoch
-        if self.op_schedule is not None:
-            self.epoch_ops = self.op_schedule.get_current_value(epoch=epoch)
-            # TODO - rather than filtering every epoch the ops should probably be separated up-front
-            self.epoch_ops = list(filter(lambda op: op.mode is None or op.mode == mode, self.epoch_ops))
-        self.state_dict = {"epoch": epoch, "mode": mode}
-
-    def get_signature_epochs(self) -> List[int]:
-        return self.signature_epochs
-
-
-class OpDataLoader(DataLoader):
-    dataset: OpDataset
-
-    def __init__(self, dataset: OpDataset, batch_size: int, shuffle: bool, num_workers: int, drop_last: bool):
-        # It may be tempting to overwrite the collate function so that it works directly with Tensorflow, but their
-        # default one has some memory management tricks that are difficult to replicate, and since we also support raw
-        # data loaders in the estimator we need to put the type conversion at that level anyways.
-        super().__init__(dataset,
-                         batch_size=batch_size,
-                         shuffle=shuffle,
-                         num_workers=num_workers,
-                         drop_last=drop_last,
-                         worker_init_fn=lambda _: np.random.seed())
-        self.set_epoch_and_mode(epoch=0, mode="train")
-
-    def set_epoch_and_mode(self, epoch: int, mode: str):
-        self.dataset.set_epoch_and_mode(epoch, mode)
-
-
-def get_signature_epochs(val: Union[Scheduler, Iterable[Any]]) -> List[int]:
-    signature_epochs = {0}
-    if isinstance(val, Scheduler):
-        signature_epochs.update(val.epoch_dict.keys())
-    else:
-        for elem in val:
-            if isinstance(elem, Scheduler):
-                signature_epochs.update(elem.epoch_dict.keys())
-    return sorted(signature_epochs)
 
+class Pipeline:
```

Comment:
we should keep the get_num_examples() call to compute the progress percentage

Reply:
The number of examples is complicating lots of things:
1. having number of examples won't allow us to calculate the progress, we also need the batch_size. 
2. calculating number of examples with tf.data.dataset doesn't work on all situations, then we have to accommodate different special cases.  


Users do have the epoch number and step number as a indicator of progress.

Reply:
what tf.data.dataset situations did you find that didn't work?

Reply:
when ds = ds.repeat()

when we call ds.from_generator